singular: update to 4.3.2, rebuild sagemath, networkx: update to 3.1

tornaria opened the pull request:
Updates:
 - singular to 4.3.2
 - python3-networkx to 3.1

For these updates we need to rebuild sagemath with patches backported from development sagemath.

In addition we also add patches to sagemath to support the following pkg updates:
 - matplotlib 3.7
 - ipython 8.12
 - nauty 2.8.6
 - pythran 0.12.1

After this PR, sagemath testsuite should pass 100% with current versions of dependencies in void. All patches are taken from upstream.

#### Testing the changes
- I tested the changes in this PR: **briefly**

I've been running this for a while. I'm also running sagemath 10.0.rc1 which includes all of our patches.


<!-- Note: If the build is likely to take more than 2 hours, please add ci skip tag as described in
https://github.com/void-linux/void-packages/blob/master/CONTRIBUTING.md#continuous-integration
and test at least one native build and, if supported, at least one cross build.
Ignore this section if this PR is not skipping CI.
-->
<!--
#### Local build testing
- I built this PR locally for my native architecture, (ARCH-LIBC)
- I built this PR locally for these architectures (if supported. mark crossbuilds):
  - aarch64-musl
  - armv7l
  - armv6l-musl
-->

tornaria commented:
In addition to the checks here, I've run the long testsuite locally on x86_64 (CI runs only normal tests since running long tests takes twice the time).

Note this PR includes (a rebase of) PR #43153.

tornaria commented:
@dkwo

dkwo commented on common/shlibs:
```diff
@@ -4087,11 +4087,11 @@ libumfpack.so.6 SuiteSparse-6.0.1_1
-libpolys-4.3.1.so singular-4.3.1p2_1
-libomalloc-0.9.6.so singular-4.3.1p2_1
-libsingular_resources-4.3.1.so singular-4.3.1p2_1
+libomalloc-0.9.6.so singular-4.2.1_1
```
typo?

tornaria replied:
The soname for `libomalloc` has been the same since `4.2.1_1`, while the others seem to change every version. In practice, it's probably irrelevant since there won't be any program that links only to `libomalloc` (only `sagemath` depends on `singular`). It's annoying that they change the soname every release... I wonder if it's really necessary.

dkwo replied:
I see, sorry for the confusion

dkwo commented:
4.3.2p1 is also available, fix a few minor things

tornaria commented:
> 4.3.2p1 is also available, fix a few minor things

But it's not released, or is it? It's not in here: https://www.singular.uni-kl.de/ftp/pub/Math/Singular/SOURCES/4-3-2/

dkwo commented:
hmm, it's on github only, forget about that

dkwo commented:
then it looks good to me.
do you want me to test on aarch64?

tornaria commented:
> do you want me to test on aarch64?

That would be great, and let me know anything weird. We could fix it.

Can you make a list of packages that are nocross? I know of ntl, fflas-ffpack, givaro, linbox, cysignals, sympow. It would be useful to fix those so they can be built and are available oob on void linux aarch64.

dkwo commented:
Besides the ones you mentioned, the list printed by xbps-src at the beginning includes
```
boost-devel
openblas-devel *
singular *
eclib-devel
maxima-ecl
python3-Cython
python3-cypari
python3-fpyllll
python3-pplpy
python3-primecountpy
```
(* means they are in this pr).
Moreover some of their dependants, like `python-Cython,` are being built.
I did not see `givaro`, as it is probably an indirect dependency that will be built by one of these.
For the tests, I have to wait until there's a big build in progress, otherwise it builds `icu` etc.

dkwo commented:
so basically fixing the ones you listed should fix everything :)

dkwo commented:
(rebuilt and tested flintlib as well)

tornaria commented:
> aarch64:
>     * maxima: has emacs both in host and run deps, so it's not in the repo; must build with ~sbcl, otherwise it fails @paper42

- emacs is cross: we can add an `emacs` option and disable it for cross.
- sbcl is also cross, so it should also be disabled at least for cross.

Do you mean maxima also fails when both sbcl and maxima are built natively?
 
>     * sagemath: fails to start tests with `aarch64-linux-gnu.so: undefined_symbol: fmpz_set_ZZ`

That is coming from flintlib. Are you using the cross built package or the native built one? The difference is that flintlib uses ntl if built natively and not if built cross.

What's the output of
```
$ objdump -T /usr/lib/libflint.so.17 | grep fmpz_set_ZZ
000000000009f8f0 g    DF .text	00000000000001ab  Base        fmpz_set_ZZ
000000000009faa0 g    DF .text	0000000000000005  Base        fmpz_set_ZZ_p
```
What about:
```
$ for p in /usr/lib/python3.11/site-packages/sage/libs/flint/*.so ; do echo $p ; readelf -d $p | grep libflint ; done
/usr/lib/python3.11/site-packages/sage/libs/flint/arith.so
 0x0000000000000001 (NEEDED)             Shared library: [libflint.so.17]
/usr/lib/python3.11/site-packages/sage/libs/flint/flint.so
 0x0000000000000001 (NEEDED)             Shared library: [libflint.so.17]
/usr/lib/python3.11/site-packages/sage/libs/flint/fmpz_poly.so
 0x0000000000000001 (NEEDED)             Shared library: [libflint.so.17]
/usr/lib/python3.11/site-packages/sage/libs/flint/ulong_extras.so
 0x0000000000000001 (NEEDED)             Shared library: [libflint.so.17]
```

dkwo commented:
yes, it first builds emacs, then sbcl, then gets stuck at some point in gdb.
i agree, we should add option emacs, and disable emacs,sbcl for cross.

i see, that makes sense for flint: my sage test was run before i rebuild flintlib. let me try again and report.

tornaria commented:
> yes, it first builds emacs, then sbcl, then gets stuck at some point in gdb. i agree, we should add option emacs, and disable emacs,sbcl for cross.

I tried to cross build maxima with ecl, but it doesn't seem to work. The build finishes if I place `ecl` in hostmakedepends but then the binaries produced are x86_64 and not cross... Maybe it's possible to cross build with clisp, but that won't be enough for sage (we really need maxima-ecl).

What we could do is add the emacs option and set the default for aarch64 to ~emacs,~sbcl so you still have to build natively but at least you are not forced to build emacs and sbcl... 

> 
> i see, that makes sense for flint: my sage test was run before i rebuild flintlib. let me try again and report.

dkwo commented:
I agree about maxima. For tests (with check=yes):
```
sage/graphs/generators/classical_geometries.py  # Killed due to segmentation fault
sage/graphs/strongly_regular_db.pyx  # Killed due to segmentation fault
sage/parallel/map_reduce.py  # Timed out
sage/sets/recursively_enumerated_set.pyx  #Timed out
```

tornaria commented:
I did the following modifications to sagemath:
 - added a patch to fix doctests when pythran is installed (optional dependency)
 - rebased on top of #43497 to fix doctests when python3-Sphinx is installed (optional dependency)
 - added `pythran` and `python3-Sphinx` to `checkdepends` so doctests that depend on them are run

dkwo commented:
looks good to me.
have all our patches been merged upstream at this point?

tornaria commented:
Tests fail because `python3-Babel` is not rebuilt, even if I included #43497 here.

tornaria commented:
> looks good to me. have all our patches been merged upstream at this point?

Yes, everything is included in 10.0.rc0, see: https://github.com/tornaria/void-packages/tree/sagemath-next if you want to try it (should I make a draft PR?)

dkwo commented:
> should I make a draft PR?

that'd be great

tornaria commented:
Here I added support for networkx 3.1 using the patch from https://github.com/sagemath/sage/pull/35584.

For testing, I include a commit to update networkx to 3.1. This one can be merged independently (but only after the patch for sagemath to support networkx 3.1 is merged).

@icp1994

icp1994 commented:
Nice, thanks!

tornaria commented:
> Nice, thanks!

Could you please try this branch and build the update of networkx to 3.1 to see if it's working ok for your purposes?

Maybe we can just merge this PR altogether as it is now. Otherwise I will try to get this merged without the networkx update and move the networkx update to a separate PR.

icp1994 commented:
Yes, I have been already using practically the same patch as this branch for my local use of networkx since v3 came out.

tornaria commented:
@leahneukirchen could we get this merged please?

This brings sagemath up to date with current versions of packages in void; it also updates networkx which was requested by @icp1994 but cannot be updated without the patch I added here.
